From akpm@osdl.org Tue Sep 27 01:26:51 2005
Message-Id: <200509270822.j8R8Mdmb004352@shell0.pdx.osdl.net>
Subject: PCI: ipr: Block config access during BIST
To: greg@kroah.com
Cc: benh@kernel.crashing.org, paulus@samba.org, akpm@osdl.org, brking@us.ibm.com, James.Bottomley@steeleye.com
From: akpm@osdl.org
Date: Tue, 27 Sep 2005 01:21:56 -0700


From: Brian King <brking@us.ibm.com>

IPR scsi adapter have an exposure today in that they issue BIST to the adapter
to reset the card.  If, during the time it takes to complete BIST, userspace
attempts to access PCI config space, the host bus bridge will master abort the
access since the ipr adapter does not respond on the PCI bus for a brief
period of time when running BIST.  On PPC64 hardware, this master abort
results in the host PCI bridge isolating that PCI device from the rest of the
system, making the device unusable until Linux is rebooted.  This patch makes
use of some newly added PCI layer APIs that allow for protection from
userspace accessing config space of a device in scenarios such as this.

Signed-off-by: Brian King <brking@us.ibm.com>
Cc: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Cc: James Bottomley <James.Bottomley@steeleye.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


 drivers/scsi/ipr.c |    2 ++
 1 file changed, 2 insertions(+)

--- gregkh-2.6.orig/drivers/scsi/ipr.c	2005-08-28 16:41:01.000000000 -0700
+++ gregkh-2.6/drivers/scsi/ipr.c	2005-09-28 15:26:54.000000000 -0700
@@ -4944,6 +4944,7 @@
 	int rc;
 
 	ENTER;
+	pci_unblock_user_cfg_access(ioa_cfg->pdev);
 	rc = pci_restore_state(ioa_cfg->pdev);
 
 	if (rc != PCIBIOS_SUCCESSFUL) {
@@ -4998,6 +4999,7 @@
 	int rc;
 
 	ENTER;
+	pci_block_user_cfg_access(ioa_cfg->pdev);
 	rc = pci_write_config_byte(ioa_cfg->pdev, PCI_BIST, PCI_BIST_START);
 
 	if (rc != PCIBIOS_SUCCESSFUL) {
